### PR TITLE
Dutch Auction Wrapper

### PR DESCRIPTION
**Pull request by hysz**
## Description

This is a wrapper for the Dutch Auction contract. It uses the same API as the contract and includes a convenience function for encoding dutch auction asset data. 

The wrapper can currently be used on testnets. Once it has been deployed to mainnet and pushed to our ganache snapshot these addresses will be added.

## Checklist:

*   [ ] Prefix PR title with `[WIP]` if necessary.
*   [x] Add tests to cover changes as needed.
*   [ ] Update documentation as needed.
*   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    /**
+     * Dutch auction details are encoded with the asset data for a 0x order. This function produces a hex
+     * encoded assetData string, containing information both about the asset being traded and the
+     * dutch auction; which is usable in the makerAssetData or takerAssetData fields in a 0x order.
+     * @param assetData Hex encoded assetData string for the asset being auctioned.
+     * @param beginTimeSeconds Begin time of the dutch auction.
+     * @param beginAmount Starting amount being sold in the dutch auction.
+     * @return The hex encoded assetData string.
+     */
+    public static encodeDutchAuctionAssetData(assetData: string, beginTimeSeconds: BigNumber, beginAmount: BigNumber): string {
```
Comment (dekz) [inline packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts]: What did you think about moving this to assetData utils?

File: packages/contract-wrappers/test/dutch_auction_wrapper_test.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import { getLatestBlockTimestampAsync } from '@0x/contracts-test-utils';
+import { DutchAuction } from '@0x/contract-artifacts';
+import { DutchAuctionWrapper } from '../src/contract_wrappers/dutch_auction_wrapper';
+
+chaiSetup.configure();
+const expect = chai.expect;
+const blockchainLifecycle = new BlockchainLifecycle(web3Wrapper);
+
+// tslint:disable:custom-no-magic-numbers
+describe.only('DutchAuctionWrapper', () => {
```
Comment (dekz) [inline packages/contract-wrappers/test/dutch_auction_wrapper_test.ts]: only

File: packages/types/src/index.ts [hunk 3]
```diff
@@ -659,3 +659,12 @@ export interface SimpleEvmOutput {
 export interface SimpleEvmBytecodeOutput {
     object: string;
 }
+
+export interface DutchAuctionDetails {
```
Comment (dekz) [inline packages/types/src/index.ts]: Need a few changelog entries for:
* Types
* Migrations
* Contract wrappers
* Contract artifacts

File: packages/contract-addresses/src/index.ts [hunk 4]
```diff
@@ -39,6 +41,7 @@ const networkToAddresses: { [networkId: number]: ContractAddresses } = {
         assetProxyOwner: '0xf5fa5b5fed2727a0e44ac67f6772e97977aa358b',
         forwarder: '0x2240dab907db71e64d3e0dba4800c83b5c502d4e',
         orderValidator: '0x90431a90516ab49af23a0530e04e8c7836e7122f',
+        dutchAuction: '0x',
```
Comment (dekz) [inline packages/contract-addresses/src/index.ts]: Will deploy these shortly

File: packages/contract-wrappers/src/contract_wrappers.ts [hunk 5]
```diff
@@ -149,7 +149,7 @@ export class ContractWrappers {
         this.dutchAuction = new DutchAuctionWrapper(
             this._web3Wrapper,
             config.networkId,
-            contractAddresses.orderValidator,
+            contractAddresses.dutchAuction,
```
Comment (dekz) [inline packages/contract-wrappers/src/contract_wrappers.ts]: 😅

Comment (hysz) [reply packages/contract-wrappers/src/contract_wrappers.ts]: :')

File: packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * Dutch auction details are encoded with the asset data for a 0x order. This function produces a hex
+     * encoded assetData string, containing information both about the asset being traded and the
+     * dutch auction; which is usable in the makerAssetData or takerAssetData fields in a 0x order.
+     * @param assetData Hex encoded assetData string for the asset being auctioned.
+     * @param beginTimeSeconds Begin time of the dutch auction.
+     * @param beginAmount Starting amount being sold in the dutch auction.
+     * @return The hex encoded assetData string.
+     */
+    public static encodeDutchAuctionAssetData(
```
Comment (abandeali1) [inline packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts]: I believe we normally order static functions above the constructor.

File: packages/contract-wrappers/package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@0x/contract-wrappers",
-    "version": "4.1.3",
+    "version": "4.2.0",
```
Comment (abandeali1) [inline packages/contract-wrappers/package.json]: Do we ever need to manually update versions in package.json?

File: packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+     * @param sellOrder The Seller's order. This order is for the lowest amount (at the end of the auction).
+     * @return The dutch auction details.
+     */
+    public async getAuctionDetailsAsync(sellOrder: SignedOrder): Promise<DutchAuctionDetails> {
+        // type assertions
+        assert.doesConformToSchema('sellOrder', sellOrder, schemas.signedOrderSchema);
+        // get contract
+        const dutchAuctionInstance = await this._getDutchAuctionContractAsync();
+        // call contract
+        const afterAuctionDetails = await dutchAuctionInstance.getAuctionDetails.callAsync(sellOrder);
```
Comment (abandeali1) [inline packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts]: I'm a little confused by this variable name. What does "after" refer to here?

File: packages/contract-wrappers/package.json (hunk 7 above)
Comment (hysz) [reply packages/contract-wrappers/package.json]: Just double-checked and we do not. Removed.

File: packages/contract-addresses/src/index.ts [hunk 9]
```diff
@@ -70,6 +78,8 @@ const networkToAddresses: { [networkId: number]: ContractAddresses } = {
         assetProxyOwner: '0x34d402f14d58e001d8efbe6585051bf9706aa064',
         forwarder: '0xb69e673309512a9d726f87304c6984054f87a93b',
         orderValidator: '0xe86bb98fcf9bff3512c74589b78fb168200cc546',
+        // @todo hysz/dekz: Add address once deployed.
+        dutchAuction: NULL_ADDRESS,
```
[the remark this answers is not in the document]
Comment (hysz) [reply packages/contract-addresses/src/index.ts]: Ahh nice -- I thought it was a mainnet snappy. Ty 🙏

**Comment by hysz** [comment]
Rebased against `development`. Hopefully this will fix the `build-website` and we can merge.

**Pull request merged**